### PR TITLE
feat(pbts): Update consensus params to include pbts enabled

### LINKED ISSUE #2197: Introduce consensus param to enable PBTS (in genesis, in updates, etc), along with all validation to update it

### PR DESCRIPTION
**Pull request by glnro**
Closes #2197 

Will need to coordinate with @lasarojc to integrate this with #2205 

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [ ] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: proto/cometbft/types/v1/params.proto [hunk 1]
```diff
@@ -107,3 +108,13 @@ message SynchronyParams {
   google.protobuf.Duration precision = 2
```
Comment (sergio-mena) [inline proto/cometbft/types/v1/params.proto:108]: please prepend `pbts_` in both params here

File: proto/cometbft/types/v1/params.proto [hunk 2]
```diff
@@ -107,3 +108,13 @@ message SynchronyParams {
...
       [(gogoproto.stdduration) = true];
 }
+
+// PBTSParams configure how CometBFT calculates block time. BFT Time is enabled by default.
+// By enabling this parameter (setting a valid non zero value), CometBFT will forever shift to using
+// Proposer Based Timestamps. For more information see
+// https://github.com/cometbft/cometbft/tree/main/spec/consensus/proposer-based-timestamp
+message PBTSParams{
```
Comment (sergio-mena) [inline proto/cometbft/types/v1/params.proto:116]:
```suggestion
message PbtsParams{
```

Comment (cason) [reply proto/cometbft/types/v1/params.proto:116]: How about `BlockTimeParams`?

File: proto/cometbft/types/v1/params.proto [hunk 3]
```diff
@@ -17,6 +17,7 @@ message ConsensusParams {
   VersionParams   version   = 4;
   ABCIParams      abci      = 5;
   SynchronyParams synchrony = 6;
+  PBTSParams pbts = 7;
```
Comment (sergio-mena) [inline proto/cometbft/types/v1/params.proto:20]:
```suggestion
  PBTSParams      pbts = 7;
```

File: types/params.go [hunk 4]
```diff
@@ -83,6 +84,22 @@ func (a ABCIParams) VoteExtensionsEnabled(h int64) bool {
 	return a.VoteExtensionsEnableHeight <= h
 }
 
+type PBTSParams struct {
```
Comment (sergio-mena) [inline types/params.go:87]:
```suggestion
type PbtsParams struct {
```

File: types/params.go [hunk 5]
```diff
@@ -83,6 +84,22 @@ func (a ABCIParams) VoteExtensionsEnabled(h int64) bool {
...
+	PBTSEnableHeight int64 `json:"pbts_enable_height"`
```
Comment (sergio-mena) [inline types/params.go:88]:
```suggestion
	PbtsEnableHeight int64 `json:"pbts_enable_height"`
```

File: types/params.go [hunk 6]
```diff
@@ -83,6 +84,22 @@ func (a ABCIParams) VoteExtensionsEnabled(h int64) bool {
...
+}
+
+// PBTSEnabled returns true if PBTS are enabled at height h
+// and false otherwise.
+func (p PBTSParams) PBTSEnabled(h int64) bool {
```
Comment (sergio-mena) [inline types/params.go:93]:
```suggestion
func (p PbtsParams) PbtsEnabled(h int64) bool {
```

File: types/params.go [hunk 7]
```diff
@@ -101,6 +118,7 @@ func DefaultConsensusParams() *ConsensusParams {
 		Version:   DefaultVersionParams(),
 		ABCI:      DefaultABCIParams(),
 		Synchrony: DefaultSynchronyParams(),
+		PBTS:      DefaultPBTSParams(),
```
Comment (sergio-mena) [inline types/params.go:121]:
```suggestion
		Pbts:      DefaultPbtsParams(),
```

File: types/params.go [hunk 8]
```diff
@@ -151,6 +169,13 @@ func DefaultSynchronyParams() SynchronyParams {
 	}
 }
 
+// Disabled by default.
+func DefaultPBTSParams() PBTSParams {
```
Comment (sergio-mena) [inline types/params.go:173]:
```suggestion
func DefaultPbtsParams() PbtsParams {
```

File: types/params.go [hunk 9]
```diff
@@ -296,6 +341,19 @@ func (params ConsensusParams) ValidateUpdate(updated *cmtproto.ConsensusParams,
...
+	if params.PBTS.PBTSEnableHeight != 0 {
+		return errors.New("pbts already enabled")
+	}
+
+	if updated.Pbts.PbtsEnableHeight <= h {
+		return fmt.Errorf("PbtsEnableHeight cannot be updated to a past height, "+
+			"pbts enabled height: %d, current height %d",
+			params.PBTS.PBTSEnableHeight, h)
+	}
+	return nil
```
Comment (sergio-mena) [inline types/params.go:354]: This should follow the same logic as `validateUpdateABCI` above

File: types/params_test.go [hunk 10]
```diff
@@ -232,6 +232,31 @@ func TestConsensusParamsValidation(t *testing.T) {
 			}),
 			valid: false,
 		},
+		// test pbts
```
Comment (sergio-mena) [inline types/params_test.go:235]:
```suggestion
		// test pbts enable height
```

File: types/params_test.go [hunk 11]
```diff
@@ -449,6 +488,53 @@ func TestConsensusParamsUpdate_VoteExtensionsEnableHeight(t *testing.T) {
 	}
 }
 
+func TestConsensusParamsUpdate_PBTSEnableHeight(t *testing.T) {
```
Comment (sergio-mena) [inline types/params_test.go:491]: Can you introduce an exhaustive UT (like `TestConsensusParamsUpdate_VoteExtensionsEnableHeight` above)?
Or maybe you can reuse `TestConsensusParamsUpdate_VoteExtensionsEnableHeight` to also test PBTS enable height, as the logic should be the same

File: internal/test/config.go [hunk 12]
```diff
@@ -69,7 +69,11 @@ var testGenesisFmt = `{
 		"abci": {
 			"vote_extensions_enable_height": "0"
 		},
-		"version": {}
+		"version": {},
+		"feature": {
+			"vote_extensions_enable_height": "0",
```
Comment (sergio-mena) [inline internal/test/config.go:74]: Shouldn't you do this when you move over it from ABCI?

File: types/params.go [hunk 13]
```diff
@@ -244,8 +249,9 @@ func (params ConsensusParams) ValidateBasic() error {
 			params.Synchrony.Precision)
 	}
 
-	if params.PBTS.PBTSEnableHeight < 0 {
-		return fmt.Errorf("PBTS.PBTSEnableHeight must not be negative. Got: %d", params.PBTS.PBTSEnableHeight)
+	// TODO: VE move.
+	if *params.Feature.PbtsEnableHeight < 0 {
```
Comment (sergio-mena) [inline types/params.go]: Can it be `nil`?

**Review by sergio-mena — APPROVED** [review]
🚀

File: proto/cometbft/types/v1/params.proto [hunk 14]
```diff
@@ -107,3 +110,24 @@ message SynchronyParams {
...
+  // subsequent heights, precommit messages that do not contain valid extension data
+  // will be considered invalid. Prior to this height, vote extensions will not
+  // be used or accepted by validators on the network.
+  //
+  // Once enabled, vote extensions will be created by the application in ExtendVote,
+  // passed to the application for validation in VerifyVoteExtension and given
+  // to the application to use when proposing a block during PrepareProposal.
+  google.protobuf.Int64Value vote_extensions_enable_height = 1 [(gogoproto.nullable) = true];
+
+  // PBTSParams configure how CometBFT calculates block time. BFT Time is enabled by default.
```
Comment (sergio-mena) [inline proto/cometbft/types/v1/params.proto:126]:
```suggestion
  // pbts_enable_height configures how CometBFT calculates block time. BFT Time is enabled by default.
```

Comment (glnro) [reply proto/cometbft/types/v1/params.proto:127]: Vetoing as this is consistent with other proto comments

File: types/params.go [hunk 15]
```diff
@@ -83,6 +88,23 @@ func (a ABCIParams) VoteExtensionsEnabled(h int64) bool {
 	return a.VoteExtensionsEnableHeight <= h
 }
 
+// FeatureParams configure parameters of different features of CometBFT.
+type FeatureParams struct {
+	VoteExtensionsEnableHeight *int64 `json:"vote_extensions_enable_height"`
```
Comment (cason) [inline types/params.go:93]: Shouldn't it be a `types.Int64Value`?

Comment (cason) [reply types/params.go:93]: Moreover, shouldn't we add the ABCI version after?

File: internal/test/config.go [hunk 16]
```diff
@@ -69,8 +69,12 @@ var testGenesisFmt = `{
 		"abci": {
 			"vote_extensions_enable_height": "0"
 		},
-		"version": {}
-	},
+		"version": {},
+		"feature": {
+			"vote_extensions_enable_height": "0",
+			"pbts_enable_height": "0"
```
Comment (cason) [inline internal/test/config.go:75]: I wonder whether this should be set to 1 here, meaning PBTS enabled?

File: types/genesis_test.go [hunk 17]
```diff
@@ -74,10 +74,12 @@ func TestBasicGenesisDoc(t *testing.T) {
 				"synchrony":  {"precision": "1", "message_delay": "10"},
 				"validator": {"pub_key_types":["ed25519"]},
 				"block": {"max_bytes": "100"},
-				"evidence": {"max_age_num_blocks": "100", "max_age_duration": "10"}
+				"evidence": {"max_age_num_blocks": "100", "max_age_duration": "10"},
+				"feature": {"vote_extension_enable_height": "0", "pbts_enable_height": "0"}
```
Comment (cason) [inline types/genesis_test.go:78]: Same here. Shouldn't we default to PBTS at this point?

File: types/params.go [hunk 18]
```diff
@@ -83,6 +88,23 @@ func (a ABCIParams) VoteExtensionsEnabled(h int64) bool {
...
+	PbtsEnableHeight           *int64 `json:"pbts_enable_height"`
+}
+
+// PbtsEnabled returns true if PBTS are enabled at height h and false otherwise.
+func (p FeatureParams) PbtsEnabled(h int64) bool {
+	if h < 1 {
```
Comment (cason) [inline types/params.go:99]: Can this happen? The validation shouldn't allow for this.

File: types/params.go [hunk 19]
```diff
@@ -151,6 +174,16 @@ func DefaultSynchronyParams() SynchronyParams {
 	}
 }
 
+// Disabled by default.
+func DefaultFeatureParams() FeatureParams {
+	defPbtsHeight := int64(0)
```
Comment (cason) [inline types/params.go:179]: Go is not a great language in this case...

File: types/params.go [hunk 20]
```diff
@@ -358,10 +433,27 @@ func (params ConsensusParams) Update(params2 *cmtproto.ConsensusParams) Consensu
...
+		}
+	}
 	return res
 }
 
 func (params *ConsensusParams) ToProto() cmtproto.ConsensusParams {
+	feature := cmtproto.FeatureParams{}
+	if params.Feature.PbtsEnableHeight != nil {
+		feature.PbtsEnableHeight = &gogo.Int64Value{}
+		feature.PbtsEnableHeight.Value = *params.Feature.PbtsEnableHeight
```
Comment (cason) [inline types/params.go:450]: Go not being nice again. No better way to initialize a `Int64Value`?

File: types/params.go [hunk 21]
```diff
@@ -417,5 +510,13 @@ func ConsensusParamsFromProto(pbParams cmtproto.ConsensusParams) ConsensusParams
 			c.Synchrony.Precision = *pbParams.Synchrony.GetPrecision()
 		}
 	}
+	if pbParams.Feature != nil {
+		if pbParams.Feature.PbtsEnableHeight != nil {
+			c.Feature.PbtsEnableHeight = &pbParams.Feature.PbtsEnableHeight.Value
```
Comment (cason) [inline types/params.go:515]: Shouldn't we use getters here?

**Review by cason — COMMENTED** [review]
Some late comments here.

File: internal/test/config.go (hunk 16 above)
Comment (lasarojc) [reply internal/test/config.go:75]: maybe as a second test case

File: proto/cometbft/types/v1/params.proto [hunk 22]
```diff
@@ -5,18 +5,21 @@ option go_package = "github.com/cometbft/cometbft/api/cometbft/types/v1";
...
 
 // ConsensusParams contains consensus critical parameters that determine the
 // validity of blocks.
+// TODO: move ABCI to FeatureParams
 message ConsensusParams {
   BlockParams     block     = 1;
   EvidenceParams  evidence  = 2;
   ValidatorParams validator = 3;
   VersionParams   version   = 4;
   ABCIParams      abci      = 5;
```
Comment (lasarojc) [inline proto/cometbft/types/v1/params.proto:20]: we need to get rid of this.

File: types/genesis_test.go (hunk 17 above)
Comment (lasarojc) [reply types/genesis_test.go:78]: Good the same approach as for VE, not on by default.
We should probably have test cases for both situations from Genesis, though

File: types/params.go (hunk 15 above)
Comment (lasarojc) [reply types/params.go:93]: We use the Int64Value only to make protobuffer entries nullable. The domain types are kept "simpler".

File: types/params.go (hunk 18 above)
Comment (lasarojc) [reply types/params.go:99]: It should never happen.

File: types/params.go (hunk 20 above)
Comment (lasarojc) [reply types/params.go:450]: Agreed. Couldn't find any.

File: types/genesis_test.go (hunk 17 above)
Comment (cason) [reply types/genesis_test.go:78]: Will address in [#2324](https://github.com/cometbft/cometbft/issues/2324)

File: internal/test/config.go (hunk 16 above)
Comment (cason) [reply internal/test/config.go:75]: Will address in [#2324](https://github.com/cometbft/cometbft/issues/2324)

File: types/params_test.go (hunk 11 above)
Comment (lasarojc) [reply types/params_test.go:492]: Done in [`TestConsensusParamsUpdate_EnableHeight`](https://github.com/cometbft/cometbft/blob/bd464ea7f8870107e15598fadc5d208d134da330/types/params_test.go#L467-L468)

File: types/params.go (hunk 13 above)
Comment (lasarojc) [reply types/params.go]: No longer applicable.

**Pull request merged**